src/librbd/internal.cc

```diff
@@ -2078,7 +2082,66 @@ void filter_out_mirror_watchers(ImageCtx *ictx,
     bufferlist *m_bl;
   };
 
-  int copy(ImageCtx *src, ImageCtx *dest, ProgressContext &prog_ctx)
+
+  class C_CopyReadSparse : public Context {
```

Comment:
Nit: I'd prefer to see ```C_CopyRead``` and ```C_CopyReadSparse``` combined -- update ```calc_sparse_extent``` to handle the "sparse_size == 0" case